curp/src/server/cmd_worker.rs

```diff
@@ -0,0 +1,536 @@
+    async fn recv(&self) -> Result<(ExeMsg<C>, DoneNotifier), flume::RecvError>;
+}
+
+/// Cmd after sync recv interface
+#[cfg_attr(test, automock)]
+#[async_trait]
+trait CmdAsReceiverInterface<C: Command + 'static> {
+    /// Recv execute msg and done notifier
+    async fn recv(&self) -> Result<(AsMsg<C>, DoneNotifier), flume::RecvError>;
+}
+
+#[async_trait]
+impl<C: Command + 'static> CmdExeReceiverInterface<C> for CmdExeReceiver<C> {
+    /// Recv execute msg and done notifier
+    async fn recv(&self) -> Result<(ExeMsg<C>, DoneNotifier), flume::RecvError> {
+        self.0.recv_async().await
+    }
+}
+
+#[async_trait]
+impl<C: Command + 'static> CmdAsReceiverInterface<C> for CmdAsReceiver<C> {
+    /// Recv execute msg and done notifier
+    async fn recv(&self) -> Result<(AsMsg<C>, DoneNotifier), flume::RecvError> {
+        self.0.recv_async().await
+    }
+}
+
+/// Create a channel to send cmds to background cmd execute workers. The channel guarantees the execution order and that after sync is called after execution completes
+pub(super) fn cmd_exe_channel<C: Command + 'static>(
+) -> (CmdExeSender<C>, CmdExeReceiver<C>, CmdAsReceiver<C>) {
+    let (exe_tx, rx) = conflict_checked_mpmc::channel();
+    let (as_tx, as_rx) = conflict_checked_mpmc::channel();
+    (
+        CmdExeSender { exe_tx, as_tx },
+        CmdExeReceiver(rx),
+        CmdAsReceiver(as_rx),
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use parking_lot::{Mutex, RwLock};
+    use tokio::{sync::mpsc, time::Instant};
+    use tracing_test::traced_test;
+
+    use super::*;
+    use crate::{
+        server::{cmd_board::CommandBoard, spec_pool::SpeculativePool},
+        test_utils::{
+            sleep_millis, sleep_secs,
+            test_cmd::{TestCE, TestCommand},
+        },
+    };
+
+    // This should happen in fast path in most cases
+    #[traced_test]
+    #[tokio::test]
+    async fn fast_path_normal() {
+        let (er_tx, mut er_rx) = mpsc::unbounded_channel();
+        let (as_tx, mut as_rx) = mpsc::unbounded_channel();
+        let ce = Arc::new(TestCE::new("S1".to_owned(), er_tx, as_tx));
+        let (exe_tx, exe_rx, exe_as_rx) = cmd_exe_channel::<TestCommand>();
+        let cmd_board = Arc::new(RwLock::new(CommandBoard::new()));
+        let spec_pool = Arc::new(Mutex::new(SpeculativePool::new()));
+        tokio::spawn(execute_worker(
+            exe_rx,
+            Arc::clone(&cmd_board),
+            spec_pool,
+            Arc::clone(&ce),
+        ));
+        tokio::spawn(after_sync_worker(exe_as_rx, cmd_board, Arc::clone(&ce)));
+
+        let cmd = Arc::new(TestCommand::default());
+        exe_tx.send_exe(Arc::clone(&cmd));
+        assert_eq!(er_rx.recv().await.unwrap().1, vec![]);
+
+        exe_tx.send_after_sync(Arc::clone(&cmd), 1);
+        assert_eq!(as_rx.recv().await.unwrap().1, 1);
+    }
+
+    // When the execution takes more time than sync, as should be called after exe has finished
```

Comment:
If a variable name has same spell as some English words, we'd better put it in backquotes to avoid misunderstanding.

Reply:
Sure.